### PR TITLE
Search: Add a Gutenberg powered customization experience

### PR DESCRIPTION
**Pull request by jsnmoon**
EDIT: Please review #20282 instead of this PR.

#### Changes proposed in this Pull Request:
Screen Shot 2021-06-30 at 5 12 13 PM

* Adds a fully functioning Jetpack Search customization experience powered by Gutenberg codenamed "Customberg" 😉.
  * Creates a new wp-admin page at `/wp-admin/admin.php?page=jetpack-search-customize`.
  * Serves a new React application that provides a Gutenberg-styled search customization experience.
  * Basic analytics instrumented via Tracks.
  * Fully responsive design.
* Updates Jetpack Search customization links in the Jetpack dashboard to point to Customberg on the following pages:
  * At-a-glance page.
  * My Plans page.
  * Performance settings page.
* Updates a `@wordpress` build package to ensure compatibility with `@wordpress/interface`, which Customberg relies on.
  * `@wordpress/dependency-extraction-webpack-plugin`
* Updates the search app build process, adds a new `watch-search` command.

---

#### Jetpack product discussion
See pcNPJE-7J-p2 and #20128.

#### Does this pull request change what data or activity we track or use?
Yes, we add additional Tracks analytics for the initial page view and when the save button is clicked.

---

#### Testing instructions:
Apply these changes to your site with a Jetpack Search subscription.

* Testing new links:
  * Visit the At-a-glance page (`/wp-admin/admin.php?page=jetpack#/dashboard`) and locate the "Customize your Search Experience" link inside the Search card. Ensure that it links to Customberg instead of the Customizer.
  * Visit the My Plans page (`/wp-admin/admin.php?page=jetpack#/my-plan`) and locate the "Customize Search" button inside the "Instant search and filtering" card. Ensure that it links to Customberg instead of the Customizer.
  * Visit the Performance settings page (`/wp-admin/admin.php?page=jetpack#/performance`). Ensure that you've enabled Instant Search and that there's a link with the text "Customize your Search experience." Ensure that it links to Customberg instead of the Customizer.

* Testing build and dependency updates:
  * Ensure that all existing JS tests pass as expected.
  * Ensure that `pnpm build` and `pnpm watch` work as expected.
  * Ensure that `pnpm build-search` and `pnpm watch-search` work as expected.

* Testing the new Search customization experience ("Customberg"):
  * Navigate to the new customization page at `/wp-admin/admin.php?page=jetpack-search-customize`.
  * Ensure that the page renders a Gutenberg layout with a header, a sidebar to the right, and the Jetpack Search application in the center/left.
  * Ensure that you can customize the search application via the controls in the options tab within the sidebar.
  * Ensure that the "Jetpack Search" tab renders a description of Customberg and includes a link to the Customizer.
  * Try resizing the page. Ensure that the page works as expected in all viewports ranging from mobile to desktop.

### CODE REVIEW

**Comment by bluefuton** [comment]
I'm sure you've considered this, but is there any practical way to break this into smaller chunks to review @jsnmoon?

File: projects/plugins/jetpack/_inc/client/at-a-glance/search.jsx [hunk 1]
```diff
@@ -163,7 +163,7 @@ class DashSearch extends Component {
 						<Card
 							compact
 							className="jp-search-config-aag"
-							href="customize.php?autofocus[section]=jetpack_search"
+							href="admin.php?page=jetpack-search-customize"
```
Comment (bluefuton) [inline projects/plugins/jetpack/_inc/client/at-a-glance/search.jsx:166]: I'd suggest updating these links in a followup PR when we're ready to ship the final customization experience.

File: projects/plugins/jetpack/_inc/client/my-plan/my-plan-body.jsx [hunk 2]
```diff
@@ -259,7 +259,7 @@ class MyPlanBody extends React.Component {
 						</p>
 						<Button
 							onClick={ this.handleButtonClickForTracking( 'view_search_customizer' ) }
-							href={ this.props.siteAdminUrl + 'customize.php?autofocus[section]=jetpack_search' }
+							href={ this.props.siteAdminUrl + 'admin.php?page=jetpack-search-customize' }
```
Comment (bluefuton) [inline projects/plugins/jetpack/_inc/client/my-plan/my-plan-body.jsx:262]: I'd suggest updating these links in a followup PR when we're ready to ship the final customization experience.

File: projects/plugins/jetpack/_inc/client/performance/search.jsx [hunk 3]
```diff
@@ -138,7 +138,7 @@ function Search( props ) {
 				<Card
 					className="jp-settings-card__configure-link"
 					compact
-					href="customize.php?autofocus[section]=jetpack_search"
+					href="admin.php?page=jetpack-search-customize"
```
Comment (bluefuton) [inline projects/plugins/jetpack/_inc/client/performance/search.jsx:141]: I'd suggest updating these links in a followup PR when we're ready to ship the final customization experience.

**Comment by bluefuton** [comment]
Can we add a README describing the general approach, and React vs Preact interoperability etc?

File: projects/plugins/jetpack/modules/search/instant-search/lib/tracks.js [hunk 4]
```diff
@@ -10,27 +10,32 @@ export function disableAnalytics() {
 
 /**
  * Initalizes Tracks.
+ *
+ * @param {boolean} forceEnableAnalytcs - Forcibly enable analytics, ignoring the isAnalyticsEnabled flag.
  */
-export function initializeTracks() {
-	if ( isAnalyticsEnabled ) {
+export function initializeTracks( forceEnableAnalytcs = false ) {
```
Comment (bluefuton) [inline projects/plugins/jetpack/modules/search/instant-search/lib/tracks.js:16]: `forceEnableAnalytcs` -> `forceEnableAnalytics`?

File: projects/plugins/jetpack/modules/search/customberg/lib/_base-styles.scss [hunk 5]
```diff
@@ -0,0 +1,29 @@
+@import '~@wordpress/base-styles/z-index';
+
+// Fill in missing z-layer values from a more recent @wordpress/base-styles version.
+// TODO: Remove once @wordpress/base-styles is updated.
```
Comment (bluefuton) [inline projects/plugins/jetpack/modules/search/customberg/lib/_base-styles.scss:4]: Are we waiting on a new release of base-styles?

**Comment by jsnmoon** [comment]
> I'm sure you've considered this, but is there any practical way to break this into smaller chunks to review @jsnmoon? (@bluefuton)

Sure, I'll look into it again; this is the result of breaking this into a smaller chunk as-is 😆

(It might be worth noting that the pnpm lock file accounts for 300+ LOC.)

File: projects/plugins/jetpack/modules/search/customberg/lib/_base-styles.scss (hunk 5 above)
Comment (jsnmoon) [reply projects/plugins/jetpack/modules/search/customberg/lib/_base-styles.scss:4]: Upgrading `base-styles` will break other parts of Jetpack, so I decided against performing the upgrade purely for our sake.

**Comment by jsnmoon** [comment]
See #20282 instead of this PR.

**Pull request closed without merging**